Strings in Ruby

pulgamecanica opened the pull request:
## What GitHub issue does this PR apply to?


Resolves #426 


## What changed and why?

   I implemented the Strings module for the Ruby 3.x lang 😄 
   I used the official Ruby [documentation](https://ruby-doc.org/core-3.1.2/String.html) as well as [The Well-Grounded Rubyist](https://www.manning.com/books/the-well-grounded-rubyist-third-edition) in case you want to add some references in the future

## Checklist

- [x] I claimed any associated issue(s) and they are not someone else's
- [x] I have looked at documentation to ensure I made any revisions correctly
- [x] I tested my changes locally to ensure they work
- [x] (If editing Django) I have added or edited any appropriate unit tests for my changes

pulgamecanica committed:
JSON finished not reviewed

pulgamecanica committed:
Resolved question 2 about strings

pulgamecanica committed:
Fixed small typo to show the newline char explicitly

geekygirlsarah committed:
Merge branch 'main' into issue-426-ruby-add-strings

geekygirlsarah commented on web/thesauruses/ruby/3/strings.json:
```diff
@@ -0,0 +1,416 @@
+    "import": {
+      "name": "Import the string class",
+      "code": [
+        "Built-in included in the default Modules"
```
Since this isn't actually code but is really a comment, would you mind changing `"code":` to `"comment":` ?

(same for others below)

pulgamecanica replied:
Alright! I get it, I'll change it, it makes a lot of sense! :D

pulgamecanica commented:
Hey! I wish I was more careful, I have one mistake.
I misplaced two questions and I forgot to answer one.

The questions misplaced should be one row above, and I should answer the black space that would be left

***

<img width="1017" alt="Screen Shot 2022-10-04 at 18 52 14" src="https://user-images.githubusercontent.com/28810331/193891211-d4961f6d-2fa8-43b8-9c5a-f06098711cfe.png">

***

I feel very noob r, wow, srry.

Tbh I feel like, after all, it's good to have PR's instead of committing directly...

pulgamecanica committed:
Fixed missplaced string concatenation answer

pulgamecanica commented:
Fixed misplaced section and answered the question about alpha strings.

geekygirlsarah commented on web/thesauruses/ruby/3/strings.json:
```diff
@@ -0,0 +1,416 @@
+    "create_new_string": {
+      "name": "Create new string",
+      "code": [
+        "1. varName = \"Hello World\"",
```
I also really want to keep "code" lines to be actual compilable code. So `varName = "Hello World"` is compilable, but the "1." at the beginning isn't. (I know you're using it to mark the line of possible options.) While maybe the system should number the lines (an issue I think I'm going to put in), for now, maybe remove the numberings and keep it to just working code.

(also the comments after should go in a "comment" block instead of "code" lines.)

pulgamecanica replied:
Understood! Yes, it was to make very clear the different options, I will change it and try to make it very understandable.

geekygirlsarah reviewed (changes requested):
This looks great! A couple of changes to match some best practices and I'll happily merge this in!

pulgamecanica commented:
I will read this before going any further...
[https://docs.codethesaur.us/thesaurus/](https://docs.codethesaur.us/thesaurus/)

pulgamecanica commented:
## Hello, I wanted to ask about this behavior...
### Whenever I replace "code" for "comment", then the application will display "Unknown" above the comment, this is because the application assumes that "code" exists, but I didn't declare it on the JSON scope, I only declared the "comment".
<img width="2033" alt="Screen Shot 2022-10-05 at 11 19 48" src="https://user-images.githubusercontent.com/28810331/194038561-dfa48c2f-5953-4992-94b8-61572d28b65c.png">

### There is one way to fix it, which would be to declare the "code" as an empty array on the JSON scope, but this will leave a blank space above the comment.

<img width="2002" alt="Screen Shot 2022-10-05 at 11 23 21" src="https://user-images.githubusercontent.com/28810331/194038907-7efce810-1f62-4334-89d6-64310dcc61d4.png">

***

Perhaps there should be a condition to check if "code" is empty, and if that's the case, only display the comment line, thus removing the blank space for the empty comment.

pulgamecanica commented:
I also found out why it says, "Unknown".
I think, if you agree with me, it's a nice change. Whenever the concept IS known, but the "code" is None then return None, instead of "unknown".

I already solved it, if you agree with me I would be more than happy to make this change happen.

pulgamecanica commented:
## My solution would be this:
***
### On the models, on this function:
https://github.com/codethesaurus/codethesaur.us/blob/229b52766c32ed2fbf8edcbe03dcc06460d650e6/web/models.py#L134

### I would change it to something like this:

```
def concept_code(self, concept_key):
        code = self.concept(concept_key).get("code")
        if not code:
            return None
        if isinstance(code, list):
            code = "\n".join(code)
        return code
```

> This way, whenever the "code" in the JSON object is not defined, you return None. (if not code, return None)

***

### On the views on this function :
https://github.com/codethesaurus/codethesaur.us/blob/229b52766c32ed2fbf8edcbe03dcc06460d650e6/web/views.py#L376

### I would change it to something like this:

```
def format_code_for_display(concept_key, lang):
    if lang.concept_unknown(concept_key):
        return "Unknown"
    if lang.concept_code(concept_key) is None:
        return None
    if lang.concept_implemented(concept_key):
        return highlight(
            lang.concept_code(concept_key),
            get_lexer_by_name(lang.key, startinline=True),
            HtmlFormatter()
        )
    return None

```

> So, if the concept is defined but the code is None, then you return None, instead of "Unknown"

pulgamecanica commented:
This will actually make this line work:

https://github.com/codethesaurus/codethesaur.us/blob/229b52766c32ed2fbf8edcbe03dcc06460d650e6/web/templates/concept_card.html#L5

And the div will not be created, because "code" it's returning None...

pulgamecanica committed:
Implement the 'comment' JSON variable when it's necessary

pulgamecanica commented:
#### Hello, I implemented the **" comment"**, and used the **"not-implemented": true** where I thought it was necessary.
#### I also fixed the ordered list.
***

I will open an issue regarding what I described above, I think it's better, that way we can separate these issues.
Best regards,
André

pulgamecanica committed:
Remove name field from the JSON file

pulgamecanica commented:
I think I've finished what was left for this String module, I finally deleted the "name" field from the JSON file, implemented the comments; etc as I described before. I would be more than happy to hear your feedback, and I was excited about creating a PR for the "Unkown" issue if you can have a  look at it at some point.
Cheers 😀

geekygirlsarah commented:
> I feel very noob r, wow, srry.
> 
> Tbh I feel like, after all, it's good to have PR's instead of committing directly...

Don't feel too bad. I've definitely committed stuff that broke the project before, lol!

Also a LOT of people have submitted stuff that's not exactly right. Probably at least half! So don't feel bad. It's why the review process and the GitHub Actions are there. To help validate things before they go in.

geekygirlsarah commented:
I'm starting to see a bit better why #558 is there. I'll think on that.

Want to wait on the result of #558, or want to just change "comment" back to "code" and I can pull it in now?

pulgamecanica commented:
I can change it back to code and save this JSON for when it's resolved.

geekygirlsarah commented:
That works for me!

pulgamecanica committed:
prevent Unkown meassage

pulgamecanica commented:
I believe it's done :D

pulgamecanica committed:
Merge branch 'main' into issue-426-ruby-add-strings

geekygirlsarah commented:
This works for me! Thank you for this addition, and your first contribution to CT! 🎉 

If you are participating in [Hacktoberfest](https://hacktoberfest.com/), don't forget to register so that this PR counts towards that.
